// material-ui-image/lib/components/Image/Image.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _CircularProgress = require('material-ui/CircularProgress');

var _CircularProgress2 = _interopRequireDefault(_CircularProgress);

var _brokenImage = require('material-ui/svg-icons/image/broken-image');

var _brokenImage2 = _interopRequireDefault(_brokenImage);

var _colors = require('material-ui/styles/colors');

var colors = _interopRequireWildcard(_colors);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function getRandomColor() {
  var colorNames = Object.keys(colors);
  return colors[colorNames[Math.floor(colorNames.length * Math.random())]];
}

/**
 * Images are ugly until they're loaded. Materialize it with material image! It will show a random material color as background and a loading animation until it's fully loaded.
 * @see [Image loading patterns](https://material.io/guidelines/patterns/loading-images.html)
 */

var Image = function (_Component) {
  _inherits(Image, _Component);

  function Image(props) {
    _classCallCheck(this, Image);

    var _this = _possibleConstructorReturn(this, (Image.__proto__ || Object.getPrototypeOf(Image)).call(this, props));

    _this.state = {
      color: getRandomColor()
    };
    return _this;
  }

  _createClass(Image, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (nextProps.src !== this.props.src) {
        this.setState({
          color: getRandomColor()
        });
      }
    }
  }, {
    key: 'getStyles',
    value: function getStyles() {
      var _props = this.props,
          aspectRatio = _props.aspectRatio,
          color = _props.color,
          errorSize = _props.errorSize,
          style = _props.style;


      var styles = {
        root: {
          backgroundColor: color || this.state.color,
          paddingTop: 'calc(1 / ' + aspectRatio + ' * 100%)',
          position: 'relative'
        },
        image: {
          width: '100%',
          height: '100%',
          opacity: !this.state.imageLoaded ? 0 : 1,
          animation: !this.state.imageLoaded ? '' : 'filter-animation 1s',
          position: 'absolute',
          top: 0,
          left: 0
        },
        errorIcon: {
          width: errorSize,
          height: errorSize
        }
      };

      return styles;
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var styles = this.getStyles();

      var _props2 = this.props,
          aspectRatio = _props2.aspectRatio,
          color = _props2.color,
          disableError = _props2.disableError,
          disableSpinner = _props2.disableSpinner,
          errorSize = _props2.errorSize,
          imageStyle = _props2.imageStyle,
          style = _props2.style,
          loadingSize = _props2.loadingSize,
          loadingStyle = _props2.loadingStyle,
          onTouchTap = _props2.onTouchTap,
          image = _objectWithoutProperties(_props2, ['aspectRatio', 'color', 'disableError', 'disableSpinner', 'errorSize', 'imageStyle', 'style', 'loadingSize', 'loadingStyle', 'onTouchTap']);

      return _react2.default.createElement(
        'div',
        {
          style: _extends({}, styles.root, style),
          onTouchTap: onTouchTap
        },
        image.src && !this.state.imageError ? _react2.default.createElement('img', _extends({}, image, {
          style: _extends({}, styles.image, imageStyle),
          onLoad: function onLoad() {
            return _this2.setState({ imageLoaded: true });
          },
          onError: function onError() {
            return _this2.setState({ imageError: true });
          }
        })) : _react2.default.createElement(
          'div',
          { style: {
              width: '100%',
              height: '100%',
              position: 'absolute',
              top: 0,
              left: 0,
              display: 'flex',
              alignItems: 'center',
              justifyContent: 'center'
            } },
          !disableSpinner && !this.state.imageLoaded && !this.state.imageError && _react2.default.createElement(_CircularProgress2.default, {
            size: loadingSize,
            style: _extends({}, loadingStyle)
          }),
          !disableError && this.state.imageError && _react2.default.createElement(_brokenImage2.default, {
            color: colors.grey300,
            style: styles.errorIcon
          })
        )
      );
    }
  }]);

  return Image;
}(_react.Component);

exports.default = Image;


Image.defaultProps = {
  aspectRatio: 1,
  disableError: false,
  disableSpinner: false,
  errorSize: 48,
  loadingSize: 40
};

Image.propTypes = {
  /** Specifies the URL of an image. */
  src: _propTypes2.default.string.isRequired,
  /** Override aspect ratio. */
  aspectRatio: _propTypes2.default.number,
  /** Override the background color. */
  color: _propTypes2.default.string,
  /** Disables the error icon if set to true. */
  disableError: _propTypes2.default.bool,
  /** Disables the loading spinner if set to true. */
  disableSpinner: _propTypes2.default.bool,
  /** Set the size of the error icon. */
  errorSize: _propTypes2.default.number,
  /** Override the inline-styles of the image. */
  imageStyle: _propTypes2.default.object,
  /** Set the size of the refresh indicator. */
  loadingSize: _propTypes2.default.number,
  /** Override the inline-styles of the refresh indicator. */
  loadingStyle: _propTypes2.default.object,
  /** Fired when the user clicks on the image happened. */
  onTouchTap: _propTypes2.default.func,
  /** Override the inline-styles of the root element. */
  style: _propTypes2.default.object
};
